src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/MethodBodyScanner.cs

```diff
@@ -1237,6 +1226,14 @@ internal MultiValue DereferenceValue(MultiValue maybeReferenceValue, ValueBasicB
             ValueNodeList methodArguments = PopCallArguments(currentStack, calledMethod, callingMethodBody, isNewObj,
                                                              offset, out newObjValue);
 
+            // Multi-dimensional array access is represented as a call to a special Get method on the array (runtime provided method)
+            // We don't track multi-dimensional arrays in any way, so return unknown value.
+            if (calledMethod is ArrayMethod { Kind: ArrayMethodKind.Get })
+            {
+                currentStack.Push(new StackSlot(UnknownValue.Instance));
+                return;
+            }
```

Comment:
What problem is this fixing?

There are other kinds of magic methods that don't cast to EcmaMethod (if that's the problem). I would rather just have this bail when we need to cast (if that's what the problem is).

When I was originally doing the first DataFlow iteration, I ran it once in mode where we would run the dataflow scanner on all IL we compile just to make sure it can handle all the patterns (i.e. make it so that everything is interesting to dataflow and do a testpass either in CI or locally).

Reply:
It's not about the method being EcmaMethod - the code worked fine without this change. But it was providing different behavior as compared to linker/analyzer. Linker/analyzer doesn't see multi-dimensional array access as a method call, so the difference is when trying to get a value from multi-dimensional array:
* linker/analyzer -> Unknown value - since it's not a handled case
* AOT -> return value from a magical method

The warnings around this are "Unknown value passed into ..." and "<magical method> return value passed into ..." respectively for AOT.

If we think this is OK to have a different I'd be fine removing this and fixing the tests.

Reply:
In that case keeping it sounds like a good plan. We might want to extend it to the `Address` method.

The repro for that is:

```csharp
Type[,] types = new Type[1, 1];
ref Type t = ref types[0, 0];
t?.GetMethods();
```

For a moment I thought I would ask how Cecil surfaces it, but maybe I don't want to know. I probably knew in the past but now it's a repressed memory (it is a method call in IL).

Reply:
The Cecil's representation is actually not "bad" - it is the method call, but it's unresolvable method, so we fail to process the method call and return Unknown. In NativeAOT the typesystem actually resolves the method... that is the difference.

Reply:
The Address works by accident - it uses `ldind_ref` instruction which we don't handle so it's going to end up `Unknown` regardless of how we handle the method call. I will make the change anyway.